File: ejemplos/nerve/07-bot-discord-nerve/ejemplo.py
"""
Ejemplo: Bot básico con discord.py

Qué enseña este ejemplo:
  - Cómo configurar y arrancar un bot de Discord.
  - Cómo escuchar eventos (`on_ready`, `on_message`).
  - El uso de `async`/`await` para programación asíncrona (requerido por Discord).

Para tu reto (Bot Discord Nerve):
  En vez de solo responder con "¡Hola!", conectarás el bot de Discord con
  tu cliente Nerve. Cuando llegue un mensaje a Discord, usarás `cliente.send()`
  para mandarlo a la red Nerve, y viceversa.

Glosario:
  discord.Client()      — La clase base que maneja la conexión con los servidores de Discord.
  discord.Intents       — Los permisos que le pides a Discord. message_content es obligatorio 
                          para poder leer el texto de los mensajes.
  @client.event         — Un decorador que le dice a la librería que esa función manejará un evento.
  async def             — Declara una función asíncrona.
  await                 — Espera a que una tarea asíncrona (como enviar un mensaje a internet) termine.
"""

import discord
import os

# 1. Configurar los "intents" (permisos de qué eventos queremos recibir)
intents = discord.Intents.default()
intents.message_content = True  # ¡Importante para poder leer lo que escriben!

# 2. Inicializar el cliente del bot
client = discord.Client(intents=intents)

# 3. Definir qué pasa cuando el bot se conecta exitosamente
@client.event
async def on_ready():
    print(f'✅ Conectado exitosamente como {client.user}')

# 4. Definir qué pasa cuando alguien envía un mensaje en un canal
@client.event
async def on_message(message):
    # Ignorar mensajes del propio bot para evitar bucles infinitos
    if message.author == client.user:
        return

    # Si alguien dice "ping", el bot responde "pong"
    if message.content.lower() == 'ping':
        print(f"[{message.author}] dijo ping, respondiendo...")
        # 'await' es necesario siempre que interactuemos con la API de Discord
        await message.channel.send('pong 🏓')

def main():
    # Debes reemplazar esto con tu token real o usar una variable de entorno
    token = os.environ.get("DISCORD_TOKEN", "TU_TOKEN_AQUI")
    
    if token == "TU_TOKEN_AQUI":
        print("⚠️ ADVERTENCIA: Reemplaza 'TU_TOKEN_AQUI' con el token de tu bot de Discord.")
        print("Puedes obtenerlo en: https://discord.com/developers/applications")
        return
        
    print("Iniciando bot...")
    client.run(token)

if __name__ == "__main__":
    main()

File: ejemplos/nerve/07-bot-discord-nerve/ejemplo.rs
/**
 * Ejemplo: Bot básico con serenity
 * 
 * Qué enseña este ejemplo:
 *   - Cómo configurar e iniciar un cliente de serenity en Rust.
 *   - Implementar el trait EventHandler para manejar eventos.
 *   - El uso de "GatewayIntents" para especificar qué eventos quieres.
 * 
 * Para tu reto (Bot Discord Nerve):
 *   En vez de solo responder con "¡Hola!", conectarás el bot de Discord con
 *   tu cliente Nerve. Cuando llegue un mensaje a Discord, usarás tu cliente 
 *   Nerve para notificar a la red, y viceversa.
 * 
 * Glosario:
 *   serenity::Client::builder() — Constructor del cliente de Discord.
 *   GatewayIntents              — Sistema de permisos. Necesitas GUILD_MESSAGES y MESSAGE_CONTENT.
 *   EventHandler                — Trait que implementas para recibir eventos.
 *   ctx.say(&msg.content)       — Método conveniente para responder en el mismo canal.
 */

use serenity::async_trait;
use serenity::model::channel::Message;
use serenity::model::gateway::Ready;
use serenity::prelude::*;
use std::env;

// 1. Definir nuestra estructura Handler que manejará los eventos
struct Handler;

// 2. Implementar el trait EventHandler para nuestra estructura
#[async_trait]
impl EventHandler for Handler {
    // Evento message: Se llama cada vez que se envía un mensaje
    async fn message(&self, ctx: Context, msg: Message) {
        // Ignorar mensajes enviados por nuestro propio bot (u otros bots)
        if msg.author.bot {
            return;
        }

        // Si el mensaje es "ping", respondemos "pong"
        if msg.content.to_lowercase() == "ping" {
            println!("[{}] dijo ping, respondiendo...", msg.author.name);
            
            // Enviar un mensaje de respuesta
            if let Err(why) = msg.reply(&ctx.http, "pong 🏓").await {
                println!("Error enviando mensaje: {:?}", why);
            }
        }
    }

    // Evento ready: Se llama cuando el bot se conecta exitosamente
    async fn ready(&self, _: Context, ready: Ready) {
        println!("✅ Conectado exitosamente como {}", ready.user.name);
    }
}

#[tokio::main]
async fn main() {
    // 3. Obtener el token. Reemplaza esto o usa variables de entorno
    let mut token = env::var("DISCORD_TOKEN").unwrap_or_else(|_| "TU_TOKEN_AQUI".to_string());

    if token == "TU_TOKEN_AQUI" {
        println!("⚠️ ADVERTENCIA: Reemplaza 'TU_TOKEN_AQUI' con el token de tu bot de Discord.");
        println!("Puedes obtenerlo en: https://discord.com/developers/applications");
        return;
    }

    // 4. Configurar los Intents (permisos)
    // GUILD_MESSAGES para recibir los eventos y MESSAGE_CONTENT para leer el texto
    let intents = GatewayIntents::GUILD_MESSAGES
        | GatewayIntents::MESSAGE_CONTENT
        | GatewayIntents::GUILDS;

    // 5. Crear el cliente con el token, los intents y nuestro Handler
    let mut client = Client::builder(&token, intents)
        .event_handler(Handler)
        .await
        .expect("Error creando el cliente");

    println!("Iniciando bot...");

    // 6. Iniciar la conexión
    if let Err(why) = client.start().await {
        println!("Error en el cliente: {:?}", why);
    }
}
